---
title: "Comparing predictions with common disease variants"
author: "Kristy Mualim" 
date: "04/15/2021"
output:
    html_document:
       toc: true
---

```{css echo=FALSE}
.main-container {
  max-width: 1600px;
    margin-left: auto;
      margin-right: auto;
}
```

```{r setup, include=FALSE}
# load required functions and packages
library(here)
library(plyr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(plotly)
library(stringr)
library(cowplot)
library(gtools)
library(viridis)
library(RColorBrewer)
library(DT)
library(htmltools)
library(reactable)
library(patchwork)
# set global chunk options
knitr::opts_chunk$set(echo = FALSE, warnings=FALSE)

source(paste0(snakemake@params$codeDir, "JuicerUtilities.R"))
source(paste0(snakemake@params$codeDir, "CredibleSetTools.R"))

# get input arguments 
all_predictors = strsplit(snakemake@params$predictors, " ") %>% unlist()
enrichmentTables = strsplit(snakemake@input$enrichmentFiles, " ") %>% unlist()
#cellCategories <- read.delim(snakemake@params$cellTypeTable, header=T, sep="\t")
numCellTypesDF <- read.delim(snakemake@input$enrichedCellTypes, header=T, sep="\t")
enrichVsPp.noncoding <- read.delim(snakemake@input$enrichVsPp_noncoding, header=T, sep="\t")
pred <- snakemake@wildcards$pred
```
```{r plottingEnrichmentCode, echo=FALSE, error=FALSE, message=FALSE,  warnings=FALSE}
############ DESCRIPTION ###########
# This function takes in *.Enrichment.vsScore.tsv files and outputs enrichment bar plots across different cell categories as defined by catOrder 

cellEnrichment <- read.delim(enrichmentTables[1], sep = "\t", header=TRUE, check.names=F, stringsAsFactors=F, row.names=NULL)
cellEnrichment$Method <- all_predictors[1]
#
if (length(enrichmentTables) > 1){
	for (i in 2:length(enrichmentTables)){
		tryCatch(
                {
                        temp = read.delim(file=enrichmentTables[i], sep='\t', header=TRUE, stringsAsFactors = FALSE)
			temp$Method <- all_predictors[i]
			cellEnrichment = rbind(cellEnrichment, temp)
                },
                error=function(cond) {
                        message(paste("File does not exist::", i))
                },
                finally={
                        message(paste("Processed:",i))
                })
	}
}

# Set theme of plots
mytheme <- theme_classic() + theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1), axis.text=element_text(size=13))#, axis.text = element_text(size = 13), axis.title = element_text(size = 15))
aggregateDF <- cellEnrichment
ibdEnhancerList <- cellEnrichment[, c("CellType", "Disease", "enrichment", "enrichment.NoPromoters", "n.noPromoters.FractionOverlap", "n.FractionOverlap")]
is_method <- ibdEnhancerList$Method == pred
bdEnhancerList <- ibdEnhancerList[is_method,]
diseases <- unique(ibdEnhancerList$Disease)
formatEnrichmentBarPlot <- function(p) p + geom_boxplot(outlier.shape = NA) + ylim(0, 22) + mytheme + geom_hline(yintercept=1, linetype="dashed", color="gray") + scale_color_hue(l=50, c=90) + theme(plot.title = element_text(size=15), legend.position='none')

ibdEnhancerList$Disease <- ordered(ibdEnhancerList$Disease, levels=unique(rev(ibdEnhancerList[order(ibdEnhancerList$enrichment.NoPromoters),"Disease"])))
p1 <- ggplot(ibdEnhancerList, aes(x=Disease, y=enrichment.NoPromoters, fill=Disease)) + ylab(paste0("Enrichment\n( GWAS variants / all variants)")) + labs(title=paste0("Enhancers (excluding promoter regions) across traits")) + xlab("Trait") + ylim(0, max(ibdEnhancerList$enrichment.NoPromoters))
p1 <- p1 %>% formatEnrichmentBarPlot()
# figure caption
#ggplotly(p1)
```

```{r plottingEnrichmentCode_all, echo=FALSE, error=FALSE, message=FALSE, warnings=FALSE}
p2 <- ggplot(ibdEnhancerList, aes(x=Disease, y=enrichment, fill=Disease)) + ylab(paste0("Enrichment\n( GWAS variants / all variants)")) + labs(title=paste0("Enhancers (including promoter regions)")) + xlab("Trait") + ylim(0, max(ibdEnhancerList$enrichment))
p2 <- p2 %>% formatEnrichmentBarPlot()

#ggplotly(p2)
```
```{r plottingFractionOverlap, echo=FALSE, error=FALSE, message=FALSE, warnings=FALSE}
formatEnrichmentBarPlot <- function(p) p + geom_boxplot(outlier.shape = NA) + scale_y_continuous(limits = c(0, 0.20)) + mytheme + geom_hline(yintercept=1, linetype="dashed", color="gray") + scale_color_manual(values=diseases) + scale_color_hue(l=50, c=90) + theme(legend.position='none') 
#formatEnrichmentBarPlot <- function(p) p + geom_boxplot(outlier.shape = NA) + geom_jitter(position=position_jitter(0.2), size=3) + scale_y_continuous(limits = c(0, 0.20)) + mytheme + geom_hline(yintercept=1, linetype="dashed", color="gray") + scale_color_manual(values=diseases) + scale_color_hue(l=50, c=90) + theme(legend.position='none') 
ibdEnhancerList$Disease <- ordered(ibdEnhancerList$Disease, levels=unique(rev(ibdEnhancerList[order(ibdEnhancerList$enrichment.NoPromoters),"Disease"])))
p3 <- ggplot(ibdEnhancerList, aes(x=Disease, y=n.noPromoters.FractionOverlap, fill=Disease)) + ylab(paste0("Fraction Overlap\n( GWAS variants / all variants)")) + ggtitle(paste0("Enhancers (excluding promoter regions)")) 
p3 <- p3 %>% formatEnrichmentBarPlot()
#ggplotly(p3)
```

```{r plottingFractionOverlap_all, echo=FALSE, error=FALSE, message=FALSE, warnings=FALSE, fig.width=6, fig.height=4}
p4 <- ggplot(ibdEnhancerList, aes(x=Disease, y=n.FractionOverlap, fill = Disease)) + ylab(paste0("Fraction Overlap\n( GWAS variants / all variants)")) + ggtitle(paste0("Enhancers (including promoter regions)")) 
p4 <- p4 %>% formatEnrichmentBarPlot()
#ggplotly(p4)
```

```{r numEnrichedCellTypes, echo=FALSE, error=FALSE, message=FALSE, warnings=FALSE}

numCellTypesPlot_noPromoters <- ggplot(numCellTypesDF, aes(x=Disease, y=n, fill=Disease)) + ylab(paste0("Number of Enriched CellTypes")) + ggtitle(paste0("Enhancers (including promoter regions)")) + mytheme + geom_dotplot(binaxis='y', stackdir='center') + scale_color_manual(values=diseases) + scale_color_hue(l=50, c=90) + theme(legend.position='none') + scale_y_continuous(breaks=seq(0,max(numCellTypesDF$n),10))
numCellTypesPlot_noPromotersPlotly <- ggplotly(numCellTypesPlot_noPromoters)
```


```{r cdfEnrichment, echo=FALSE, error=FALSE, message=FALSE, warnings=FALSE}
aggregateDF = aggregateDF[order(aggregateDF$enrichment.NoPromoters),]
cdf = ggplot(aggregateDF, aes(x=enrichment.NoPromoters, col=Method)) + 
	stat_ecdf(geom = "step") + ylab('Cumulative fraction') + 
	xlab('Enrichment (GWAS variants/all common variants)') + 
	theme_classic() + 
	labs(title = "Density Plot for all enhancers (w/o promoter regions)")
```

```{r plotFractionOverlapPosProb, echo=FALSE, error=FALSE, message=FALSE, warnings=FALSE, fig.width=6, fig.height=4}
## Look at overlap of all variants with ABC enhancer as function of posterior prob (in any cell type\, not just enriched cell types)
# 04122021 KSM: set this as an input instead of hardcoded
bloodRelatedTraits <- c("Baso","Eosino","Hb","LOY","Lym","MCH","MCHC","MCV","Mono","Neutro","RBC","WBC")

# set theme
mytheme <- theme_classic() #+ theme(axis.text = element_text(size = 13), axis.title = element_text(size = 15))
trait="IBD"
p <- ggplot(subset(enrichVsPp.noncoding, nTotal >= 5 & PosteriorProb <= 0.95), aes(x=log10(PosteriorProb), y=fraction*100, group=Disease)) + geom_line(color='gray') + xlab("PIP (log10)") + ylab("% noncoding variants") + ylim(0,100) + mytheme
p <- p + geom_line(data=subset(enrichVsPp.noncoding,Disease%in%bloodRelatedTraits & nTotal >= 5 & PosteriorProb <= 0.95), aes(x=log10(PosteriorProb), y=fraction*100), color='red')
p <- p + geom_line(data=subset(enrichVsPp.noncoding,Disease=="All" & PosteriorProb <= 0.95), aes(x=log10(PosteriorProb), y=fraction*100), color='black')
p <- p + geom_line(data=subset(enrichVsPp.noncoding,Disease==trait & nTotal >= 5 & PosteriorProb <= 0.95), aes(x=log10(PosteriorProb), y=fraction*100), color='blue')
p <- p + geom_hline(yintercept=7.5, linetype='dashed', color='black')
p <- p 
```
```{r plotAggregatePR , echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.width=6, fig.height=4}
mytheme <- theme_classic() #+ theme(
                           #       axis.text = element_text(size = 13),
                           #       axis.title = element_text(size = 15))

#getPRPlot <- function(pr, baseline=NULL, xlab="Recall") {
#	pr <- pr %>% group_by(Method) %>% mutate(n=n())
#	#pr <- pr[order(pr$Method), ]
#	pr.points <- pr %>% filter(n==1)
#    	pr.lines <- pr %>% filter(n>1)
#	p <- ggplot(pr.points, aes(x=Recall, y=Precision, color=Method))
#	if (!is.null(baseline)) p <- p + geom_hline(yintercept=baseline, color='gray', linetype='dashed')
#	p <- p + geom_point(size=3)
#	#p <- p + geom_line(data=pr.lines)
#	p <- p + mytheme + coord_fixed() + xlim(0,1) + ylim(0,1) + xlab(xlab) + ylab("Precision")
#	return(p)
#}

# PLOT Aggregate Precision-Recall Curves
gp.all <- read.delim(snakemake@input$allgenePredTable, sep="\t", header=T, check.names=F, stringsAsFactors=F)
aggPR <- read.delim(snakemake@input$PRtable, sep="\t", header=T, check.names=F, stringsAsFactors=F)
nRecall <- aggPR$nRecall
aggCurve <- getPRPlot(aggPR, baseline=getPrecisionBaseline(gp.all), xlab=paste0("Recall (n=",nRecall,")"))
dev.off()
```

# Summary of `r snakemake@wildcards$pred` for all traits
This report shows summary statistics produced by the disease variant validation pipeline for predictions made by **`r snakemake@wildcards$pred`**. <br> It seeks to give you a better understanding of how your prediction method has performed across all common disease traits, looking mainly at enrichment of traits in enhancers (including and excluding promoter regions), the # of traits overlapping enhancers (including and excluding promoter regions) as well as how well your prediction method links non-coding disease variants to their respective genes. <br>
The following report is done across all traits. <br>

```{r summary,  echo=FALSE, error=FALSE, message=FALSE, warning=FALSE, fig.width=18, fig.height=10}

p1_title <- p1 + ggtitle('A') 
p3_title <- p3 + ggtitle('B') 
numCellTypesPlot_noPromoters_title <- numCellTypesPlot_noPromoters + ggtitle('C')
aggCurve_title <- aggCurve + ggtitle('D') 
cdfAgg <- cdf + ggtitle('E') 

patchwork <- p1_title / p3_title
patchwork
patchwork1 <- numCellTypesPlot_noPromoters_title / (aggCurve_title + cdfAgg)
patchwork1
```

<br>
**Fig1. `r snakemake@wildcards$pred` maps connecting fine-mapped variants to enhancer, genes and celltypes. a**, Enrichment of fine-mapped variants (PIP>10%) in `r snakemake@wildcards$pred` enhancers. **b**, Fraction of fine-mapped variants (PIP>10%) overlapping `r snakemake@wildcards$pred` enhancers. **c**, Number of enriched celltypes of fine-mapped variants (PIP>10%) **d**,  Precision–recall plot of connections between noncoding credible sets and known genes. **e**, Cumulative density plot of enrichment across all traits.  
<br>

Below are interactive plots to allow further exploration of how your prediction method has performed across all common disease traits. <br>
Summary statistics like (min/max/median/quantiles) are reported here, as well as outlier points. <br> 
The first question we ask is : <br>
## To what extent are GWAS variants enriched in predicted enhancers?
### Enrichment of GWAS variants for enhancers (including promoter regions)  and (excluding promoter regions)
These barplots calculates the enrichment for fine-mapped variants (PIP >= 10%) in `r snakemake@wildcards$pred` enhancers (left) and enhancers without promoter regions (promoter regions as defined by RefSeq) (right) across all biosamples. Variants were filtered to only include non-coding distal elements.  
<br>

```{r enrichmentplot1, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.cap=cap, fig.width=14, fig.height=6}

ptly1 <- ggplotly(p1)
ptly2 <- ggplotly(p2)
ptly1
ptly2
#subplot(ptly1, ptly2, nrows=2, titleX= TRUE, titleY = TRUE)
cap <- paste0("Fig2. Enrichment of fine-mapped variants (PIP >= 10%) in enhancers across all biosamples. \n Box plots show median (middle line) and interquartile range (boxes) and whiskers show observations less than or equal to quartiles ± 1.5× the interquartile range. Outlier points are included in black.")
```

## Enrichment tables {.tabset}
Tables with values used to derive enrichment values and # fraction overlap for enhancers (including promoter regions) and enhancers (excluding promoter regions). 
```{r doSomething, include=FALSE}
htmltools::tagList(DT::datatable(matrix()))
```

```{r dataTables, echo=FALSE, results='asis'}
pred = snakemake@wildcards$pred
# get index
index = which(all_predictors==pred)[[1]]
cat("  \n###",  all_predictors[index], "  \n")
table.current = ibdEnhancerList
cat("  \n")
table.current <- table.current %>% dplyr::select(CellType, Disease, enrichment, enrichment.NoPromoters, n.FractionOverlap, n.noPromoters.FractionOverlap)
cat(knitr::knit_print(datatable(table.current, rownames = FALSE, filter="top", 
			options=list(pageLength = 5, scrollX=T))))
cat("  \n")
cat("  \n")
#}
```

# To what extent do GWAS variants overlap predicted enhancers?
## Variant overlap of GWAS variants for enhancers (including promoter regions) (left) and enhancers (excluding promoter regions) (right)
We plot the fraction overlap of variants (PIP >= 10%) that overlap `r snakemake@wildcards$pred` enhancers with and without promoter regions (promoter regions as defined by RefSeq) across all biosamples.
```{r enrichmentplot3, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.cap=cap, fig.width=14, fig.height=6}
ptly3 <- ggplotly(p3)
ptly4 <- ggplotly(p4)
#subplot(ptly3, ptly4, nrows=2, titleX = TRUE, titleY = TRUE)
ptly3
ptly4
cap <- paste0("Fig3. Fraction overlap of fine-mapped variants (PIP >= 10%) in \n enhancers across all biosamples. Box plots show median (middle line) and interquartile range (boxes) and whiskers show observations less than or equal to quartiles ± 1.5× the interquartile range. Outliers are shown in black.")
```

# How many enriched celltypes are called across all traits? 
```{r plotEnriched, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.cap=cap, fig.height=6, fig.width=10}
numCellTypesPlot_noPromoters 
cap <- paste0("Fig 4. Number of enriched celltypes (p < 0.05) when calculating for enrichment of GWAS variants / all variants across all traits.")
```

# How does the fraction of noncoding variants that overlap a given `r snakemake@wildcards$pred` enhancer in any biosample change given the PIP threshold?
Variants were filtered to include only distal non-coding. Blue represents IBD. Red represents 12 other blood traits. Black represents the average across all traits and Gray represents the remaining 57 other traits. 
```{r enrichmentplot5, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.cap=cap, fig.height=4, fig.width=6}
ggplotly(p)
cap <- paste0("Fig5. Fraction of noncoding variants above a given PIP threshold that overlap an ", pred, " enhancer in any biosample. Black line, weighted average across traits. Traces are shown for PIP thresholds above which there are at least five variants. Dashed line, fraction of all common noncoding variants that overlap ", pred, " enhancers.")
```

# How well do predictions connect noncoding credible sets to known disease genes?
These precision-recall curves seek to plot the performance of choosing the gene with the **best** score per locus. <br>
**Precision** = fraction of identified genes corresponding to the list of known genes that affect all GWAS traits <br>
**Recall** = fraction of known genes that were identified. <br>
```{r plot6, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.cap=cap, fig.width=10, fig.height=4}
aggCurve
cap <- paste0("Fig6. Precision–recall plot of connections between noncoding credible sets and known genes, where recall is the fraction of credible sets for which the known
	      gene is identified (sensitivity) and precision is the fraction of predicted
	      genes corresponding to known genes (positive predictive value). As baseline, the heuristic of assigning each GWAS credible
	      set to the closest gene—a method that is widely used to annotate GWAS
	      loc. Simiarly, a similar approach — which selects the closest
	      transcription start site (TSS) was also added. ")
```

# Methods 
**GWAS traits and loci** <br> 
Summary statistics for IBD, Crohn’s disease and ulcerative colitis (European ancestry only)51 from https://www.ibdgenetics.org/downloads.html. We obtained fine-mapping results and summary statistics for 71 other traits based on an unpublished analysis ( Jacob Ul, M. Kanai and Hillary Finucane, unpublished data) that analysed data from the UK Biobank. (fine-mapping data are available at https://www.finucanelab.org/data). In this analysis, up to 361,194 individuals of white British ancestry with available phenotypes and variants with INFO > 0.8, minor allele frequency > 0.01%, and Hardy–Weinberg equilibrium P > 1 × 10−10 were included in the GWAS. For all traits, except where specified, we considered only the ‘noncoding credible sets’—that is, those that did not contain any variant in a coding sequence or within 10 bp of a splice site annotated in the RefGene database (downloaded from UCSC Genome Browser on 24 June 2017). 
<br>

**Defining enriched biosamples for each trait** <br> 
For a given trait, we intersected variants with PIP ≥ 10% in noncoding credible sets with `r snakemake@wildcards$pred` enhancers (or other genomic annotations). For each biosample, we calculated a P value using a binomial test comparing the fraction at which PIP ≥ 10% variants overlapped `r snakemake@wildcards$pred` enhancers with the fraction at which all common variants overlap `r snakemake@wildcards$pred` enhancers in that cell type. We calculated the latter using common variants in the 1000 Genomes Projects as described in the ‘Stratified linkage disequilibrium score regression’ section. For each trait, we defined a biosample as significantly enriched for that trait if the Bonferroni-corrected binomial P value was <0.001.
<br> 

**Comparison of enrichment of fine-mapped variants in enhancer regions and other enhancer-gene predictions** <br>
We compared the enrichment of fine-mapped variants in ABC enhancers and other enhancer definitions. We analysed
each of the previous studies from Fig 1d (summary plot) reporting cell-type specific enhancer-gene predictions.  For each of the methods below, we downloaded previous predictions of enhancer–gene links and assessed
their ability to identify disease-associated genes. For this analysis, we used the predictions from each method to overlap fine-mapped variants (PIP ≥ 10%) with enhancers in any cell type and assigned variants to the predicted gene(s).
<br>

**Calculating the number of enriched celltypes (p < 0.05)**<br>
Number of enriched celltypes where the enrichment of fine-mapped variants (PIP >= 10%) in enhancers are considered significant (p < 0.05) across all traits. 

**Enhancer-gene correlation (Activity-By-Contact)** <br>
DNase accessibility and acetylation of H3K27 are commonly used to identify enhancer elements, and are predictive of the expression of nearby genes and enhancer activity in plasmid-based reporter assays. The geometric mean of DNase-seq and H3K27ac ChIP–seq signals was used to calculate Activity. Contact component of the ABC score for E–G pairs from Hi-C data (when available) or the Average Hi-C data across 10 CellTypes, which was proven to an accurate substitute when celltype specific Hi-C data is not available. To calculate the relative effect of each element to the expression of a gene, we normalized the Activity by Contact of one element for a given gene to the sum of the Activity by Contact of other nearby elements. We included all elements within 5 Mb of the gene’s promoter in this calculation.
Reference: https://www.nature.com/articles/s41588-019-0538-0

**Enhancer-gene correlation (ChromHMM2017)** <br>
Gene expression was previously correlated with five active chromatin marks (H3K27ac, H3K9ac, H3K4me1, H3K4me2 and DNase I hypersensitivity) across 56 biosamples, and these correlation links were then used to make predictions for the predicted enhancers (regions with the ‘7Enh’ ChromHMM state) in 127 biosamples from the Roadmap Epigenome Atlas. We downloaded these predictions from www.biolchem.ucla.edu/labs/ernst/roadmaplinking and made predictions using the confidence score.
Reference: https://www.nature.com/articles/nprot.2017.124

<br>
**Enhancer-gene correlation (EpiMap)** <br>
Gene expression was previously correlated with five active chromatin marks (H3K27ac, H3K9ac, H3K4me1, H3K4me2 and H3K4me3) across 304 biosamples. A negative set of correlations for each enhancer was computed using random genes in a different chromosome. We predicted links for each biosample and ChromHMM enhancer state separately (states E7, E8, E9, E10, E11 and E15). Predictions were made by training an XGBoost classifier on the positive set of all valid links against their paired negative links, using precomputed correlations and distance to the transcription start site as features, and keeping all links with a probability above 5/7. We downloaded these predictions from https://personal.broadinstitute.org/cboix/epimap/links/. 
Reference: https://www.biorxiv.org/content/10.1101/810291v2
<br> 
